### ISSUE TITLE
Proposal: A builtin for initializing optionals from pointers

### ISSUE THREAD

**Issue by SpexGuy** [ziglang/zig#6992]
A side effect of the decision on #1831 is that this code is now UB:

```zig
const BigStruct = extern struct { ... };
var inst: ?BigStruct = null;

extern fn initBigStruct(*BigStruct) c_int;

pub fn init() !void {
    // if init fails, make sure inst is still null
    errdefer inst = null;

    // intent: allow for taking pointer to payload
    // under the new rule, `inst` is now undefined.
    inst = @as(BigStruct, undefined);

    // intent: initialize payload with C code
    // .? on undefined is UB, so this is now bad
    const rc = initBigStruct(&inst.?);

    // make sure the C library actually initialized stuff
    if (rc < 0) return error.InitFailed;
}
```

I don't think there's any way to write this with the current semantics, without either initializing the struct to a defined value or copying from a temporary into the optional.

So I'm proposing a builtin that will solve this problem: `@initPtr(opt: ?T) *T`.  initPtr accepts an lvalue of type `?T` for some type T.  It returns a pointer to the payload of the optional, but also sets the optional to an undefined value.  Writing a non-undefined value to the pointer puts the optional into a valid "set" state.  In other words, after using `@initPtr`, the programmer must guarantee that either the pointer is written to, or the optional is explicitly set to `null` or a valid value.  Until one of these things happens, the value of the optional is undefined.

This would allow the above example to be written as follows:
```zig
const BigStruct = extern struct { ... };
var inst: ?BigStruct = null;

extern fn initBigStruct(*BigStruct) c_int;

pub fn init() !void {
    // make sure we write null if the pointer is not written
    errdefer inst = null;

    // get the pointer for initialization and pass it to C code
    const rc = initBigStruct(@initPtr(inst));
    
    // return error (and trigger set to null) if the pointer wasn't written.
    if (rc < 0) return error.InitFailed;
}
```

**Comment by rohlem** [comment]
>`&inst.?`

I wasn't aware that this was previously allowed. The documentation states that `.?` is equivalent to `orelse unreachable`. I had assumed that `orelse` returns a value, not an `lvalue` in the case of `@TypeOf(rhs) == noreturn`.

Also the way the builtin function is supposed to take an `lvalue` looks a bit magical. I assume there's nothing in the way of equivalently redefining it as `@initPtr(opt: *?T) *T`, right?

(Also also nitpick: The name `initPtr` does not make me think of optional types. I'd prefer `optionalInitPtr`/`optInitPtr`, if that's not too long.)

**Comment by SpexGuy** [comment]
Yeah, the documentation is actually incorrect here, `x.?` is a direct reference to the payload but the result of `x orelse unreachable` is a temporary.  You are also technically allowed to do `x.? = y;` but I don't think that's actually useful.

> the way the builtin function is supposed to take an lvalue looks a bit magical

This is true, but there is precedent.  Builtins are different from functions in a lot of ways, but especially because they can be evaluated at compile time without the `comptime` keyword and because they can accept lvalues.  They are more like keywords.  The most notable example is `@field`, which accepts an lvalue as its first parameter and evaluates to an lvalue.  That said, I'm not really opposed to changing this one to take a pointer.

> The name

I'm not too attached to `initPtr`.  I agree, something that mentions optionals would be better.

**Comment by Hadron67** [comment]
Does this also apply to tagged unions? For example, is the following similar code using tagged union legal?
```zig
const BigStruct = extern struct { ... };
var inst: union(enum){
    Some: BigStruct,
    None,
} = .None;

extern fn initBigStruct(*BigStruct) c_int;

pub fn init() !void {
    errdefer inst = .None;

    inst = .{ .Some = undefined }; // type coercion here

    const rc = initBigStruct(&inst.Some);

    if (rc < 0) return error.InitFailed;
}
```
Since the assignment to `inst` using anonymous struct literal contains a type coercion I doubt the whole variable is set to undefined according to the new rule and thus results in UB. But it might be ok if tagged union is defined to be a struct containing the tag and payload separately, so I'm not sure about this.

**Comment by michaelbartnett** [comment]
This would be great to have.

Would there be any gotchas with pointers-to-optionals?

e.g.

```zig
// now:
fn initNonNullUndefined(ptr: *anyopaque) void {
    const undef_value: T = undefined;
    const ptr_to_optional: *?T = @alignCast(@ptrCast(ptr));
    ptr_to_optional.* = undef_value;
}

// after this is implemented:
fn initNonNullUndefined(ptr: *anyopaque) void {
    const ptr_to_optional: *?T = @alignCast(@ptrCast(ptr));
    ptr_to_optional.* = @initPtr(ptr_to_optional.*); // would assign-back be needed or if would discarding it be fine?
}
```
